server/transport.go

```diff
@@ -0,0 +1,169 @@
+//
+// Author: Timothy Chen
+
+package server
+
+import (
+	"net"
+	"net/rpc"
+	"sync"
+
+	"github.com/cockroachdb/cockroach/gossip"
+	"github.com/cockroachdb/cockroach/multiraft"
+	"github.com/cockroachdb/cockroach/proto"
+	crpc "github.com/cockroachdb/cockroach/rpc"
+	"github.com/cockroachdb/cockroach/rpc/codec"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/log"
+)
+
+const (
+	raftServiceName = "MultiRaft"
+	raftMessageName = raftServiceName + ".RaftMessage"
+)
+
+// RPCTransport handles the rpc messages for multiraft.
+type RPCTransport struct {
+	gossip    *gossip.Gossip
+	rpcServer *crpc.Server
+	mu        sync.Mutex
+	servers   map[multiraft.NodeID]multiraft.ServerInterface
+	clients   map[multiraft.NodeID]*rpc.Client
+}
+
+// NewRPCTransport creates a new RPCTransport with existing gossip and rpc server.
+func NewRPCTransport(gossip *gossip.Gossip, rpcServer *crpc.Server) (multiraft.Transport, error) {
+	t := &RPCTransport{
+		gossip:    gossip,
+		rpcServer: rpcServer,
+		servers:   make(map[multiraft.NodeID]multiraft.ServerInterface),
+		clients:   make(map[multiraft.NodeID]*rpc.Client),
+	}
+
+	err := t.rpcServer.RegisterName("MultiRaft", (*transportRPCServer)(t))
+	if err != nil {
+		return nil, err
+	}
+
+	return t, nil
+}
+
+// transportServer is a type alias to separate RPC methods
+// (which net/rpc finds via reflection) from others.
+type transportRPCServer RPCTransport
+
+// RaftMessage proxies the incoming request to the listening server interface.
+func (t *transportRPCServer) RaftMessage(protoReq *proto.RaftMessageRequest,
+	resp *proto.RaftMessageResponse) error {
+	// Convert from proto to internal formats.
+	req := &multiraft.RaftMessageRequest{GroupID: protoReq.GroupID}
+	if err := req.Message.Unmarshal(protoReq.Msg); err != nil {
+		return err
+	}
+
+	t.mu.Lock()
+	server, ok := t.servers[multiraft.NodeID(req.Message.To)]
+	t.mu.Unlock()
+
+	if ok {
+		return server.RaftMessage(req, &multiraft.RaftMessageResponse{})
+	}
+
+	return util.Errorf("Unable to proxy message to node: %d", req.Message.To)
+}
+
+// Listen registers a ServerInterface to be proxied messages.
+func (t *RPCTransport) Listen(id multiraft.NodeID, server multiraft.ServerInterface) error {
+	t.mu.Lock()
+	t.servers[id] = server
+	t.mu.Unlock()
+	return nil
+}
+
+// Stop unregisters the server id from getting messages.
+func (t *RPCTransport) Stop(id multiraft.NodeID) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	delete(t.servers, id)
+}
+
+func (t *RPCTransport) getClient(id multiraft.NodeID) (*rpc.Client, error) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	client, ok := t.clients[id]
+	if ok {
+		return client, nil
+	}
+
+	nodeIDKey := gossip.MakeNodeIDGossipKey(proto.NodeID(id))
+	info, err := t.gossip.GetInfo(nodeIDKey)
+	if info == nil || err != nil {
+		return nil, util.Errorf("Unable to lookup address for node: %d. Error: %s", id, err)
+	}
+	address := info.(net.Addr)
+
+	conn, err := net.Dial("tcp", address.String())
```

Comment:
Why aren't we using cockroach/rpc/client.go's NewClient()? That way we share the same client connection being used for everything else. It also gives us access to the IsHealthy() method, which I think we'll want to use to invalidate entries in the t.clients map, so we don't run into a situation where an IP address changes (should be rare, but it's not handled here).